index.d.ts

```diff
@@ -0,0 +1,143 @@
+       * reason to call this.
+       *
+       * If `multicastInterface` is not specified, the operating system will attempt to
+       * drop membership on all valid interfaces.
+       */
+      dropSourceSpecificMembership(sourceAddress: string, groupAddress: string, multicastInterface?: string): void;
+
+      /**
+       * events.EventEmitter
+       * 1. close
+       * 2. connect
+       * 3. error
+       * 4. listening
+       * 5. message
+       */
+      addListener(event: string, listener: (...args: any[]) => void): this;
+      addListener(event: 'close', listener: () => void): this;
+      addListener(event: 'connect', listener: () => void): this;
+      addListener(event: 'error', listener: (err: Error) => void): this;
+      addListener(event: 'listening', listener: () => void): this;
+      addListener(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
+
+      emit(event: string | symbol, ...args: any[]): boolean;
+      emit(event: 'close'): boolean;
+      emit(event: 'connect'): boolean;
+      emit(event: 'error', err: Error): boolean;
+      emit(event: 'listening'): boolean;
+      emit(event: 'message', msg: Buffer, rinfo: RemoteInfo): boolean;
+
+      on(event: string, listener: (...args: any[]) => void): this;
+      on(event: 'close', listener: () => void): this;
+      on(event: 'connect', listener: () => void): this;
+      on(event: 'error', listener: (err: Error) => void): this;
+      on(event: 'listening', listener: () => void): this;
+      on(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
+
+      once(event: string, listener: (...args: any[]) => void): this;
+      once(event: 'close', listener: () => void): this;
+      once(event: 'connect', listener: () => void): this;
+      once(event: 'error', listener: (err: Error) => void): this;
+      once(event: 'listening', listener: () => void): this;
+      once(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
+
+      prependListener(event: string, listener: (...args: any[]) => void): this;
+      prependListener(event: 'close', listener: () => void): this;
+      prependListener(event: 'connect', listener: () => void): this;
+      prependListener(event: 'error', listener: (err: Error) => void): this;
+      prependListener(event: 'listening', listener: () => void): this;
+      prependListener(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
+
+      prependOnceListener(event: string, listener: (...args: any[]) => void): this;
+      prependOnceListener(event: 'close', listener: () => void): this;
+      prependOnceListener(event: 'connect', listener: () => void): this;
+      prependOnceListener(event: 'error', listener: (err: Error) => void): this;
+      prependOnceListener(event: 'listening', listener: () => void): this;
+      prependOnceListener(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
```

Comment:
Do we actually need to add all the EventEmitter methods? It should be added to the type automatically with the extends, shouldn't it?

Reply:
I think most of these are here either:

- because dgram uses variadic functions
- because the base NodeJS.EventEmitter type just lists events as `string | Symbol` etc and in the case of dgram we want to restrict those events to `string | 'open' | 'close'` to have very narrow typing and typehinting for the dgram API

Reply:
Think this is the correct way where it lists all the different events that can be interacted with, also typing the function in a specific case (e.g the extra `msg: Buffer` argument for `'message'` events vs void for `'connect'`)

Again, just copied this from the dgram TS file. Not sure if you really want to, but if you'd like to stay as close as possible to Node's dgram API, I think copying it's type declaration would be the right thing to do.

Reply:
You are right, I agree.